text/0000-mixed-shadow-mode.md

```diff
@@ -0,0 +1,205 @@
+---
+title: Mixed Shadow Mode
+status: DRAFTED
+created_at: 2021-05-07
+updated_at: YYYY-MM-DD
+pr: https://github.com/salesforce/lwc-rfcs/pull/49
+---
+
+# Mixed Shadow Mode
+
+## Introduction
+
+LWC offers a collection of polyfills that implement Shadow DOM features. These polyfills are
+collectively referred to as synthetic Shadow DOM and are published to npm under the
+`@lwc/synthetic-shadow` package. Synthetic Shadow DOM provides a consistent experience for LWC web
+components across the Salesforce-supported browser matrix by implementing Shadow DOM semantics in
+older browsers that do not support it.
+
+Native Shadow DOM support has become more common since the start of the LWC project but applications
+continue to rely on the synthetic shadow polyfills to maintain backwards-compatibility. The removal
+of these polyfills can result in broken experiences for existing applications due to dependencies on
+certain "escape hatches" that are not possible in a native Shadow DOM environment (e.g.,
+instrumentation, styling, etc).
+
+Applications that wish to use LWC components in a native shadow context simply need to omit the
+inclusion of the `@lwc/synthetic-shadow` polyfill; however, this all-or-nothing approach is not a
+realistic option for large applications that cannot afford a rewrite.
+
+This proposal opens up an incremental migration path for applications moving to native Shadow DOM by
+enabling the usage of both native and synthetic Shadow DOM in the same application.
+
+## Detailed design
+
+Up until now, the two choices of shadow semantics have been synthetic mode and native mode. All the
+components in an application run in either synthetic mode or native mode, depending on whether the
+synthetic shadow polyfill (i.e., `@lwc/synthetic-shadow`) is applied. Mixed mode introduces the
+ability for a component to access native Shadow DOM APIs, even while the synthetic shadow polyfill
+is applied. Mixed mode should be transparent to components and the framework should guarantee this
+through integration testing, which is further discussed below.
+
+### Opting in
+
+A component can signal that it supports both synthetic and native Shadow DOM by setting a static
+property called `shadowMode` to `any`.
+
+```js
+export default class extends LightningElement {
+    static shadowMode = 'any';
+}
+```
+
+This mode will result in the whole component subtree to operate in either native shadow mode or
+synthetic shadow mode. If the browser supports Shadow DOM the shadow mode will be native. If the
+browser does not support Shadow DOM (e.g., IE11) the shadow mode will be synthetic. The test to
+determine Shadow DOM support will be to check whether `window.ShadowRoot` is defined.
+
+Setting `shadowMode` to `undefined` is equivalent to not setting the property at all. A value of
+`undefined` results in the default behavior where the shadow mode of the component depends only on
+the inclusion of the `@lwc/synthetic-shadow` polyfill.

```

Comment:
@ekashida 

```suggestion
Setting `shadowMode` to `undefined` is equivalent to not setting the property at all. A value of
`undefined` results in the default behavior where the shadow mode of the component depends only on
the inclusion of the `@lwc/synthetic-shadow` polyfill. The only valid values are `undefined` and `"any"` – any other value will throw a compiler error.
```

Reply:
Actually on second thought, I think we may need to introduce an explicit value like `"default"` in addition to `undefined` and `"any"`. The problem is here:

```js
class Super extends LightningElement {
  static shadowMode = 'any'
}

class Sub extends Super {
  static shadowMode = undefined
}
```

With `renderMode`, [we allow subclasses to inherit the `static renderMode` from their superclass](https://github.com/salesforce/lwc/blob/8ddcfc44118d7ae12b3c1dcba131ba78cc66d693/packages/%40lwc/engine-core/src/framework/def.ts#L159-L167). We also currently do this for `preferNativeShadow`.

But in this case, it's ambiguous whether `Sub` should inherit the `shadowMode` from `Super` or not.